### ISSUE TITLE
Performance improvements

### ISSUE THREAD

**Issue by foxfriends** [foxfriends/lumber#35]
After #34 performance is much better but I'm sure there can still be more. It's just not as critical anymore. Maybe some `Rc` can be added to patterns (in place of boxes? in other places?) to speed up cloning even further.

**Comment by foxfriends** [comment]
Based on some profiling, it seems like most of the time currently is spent doing `Binding::transfer_from`, probably because there is no sharing between the two bindings, so it takes a long time to clone everything. Adding `Rc`s to the `Pattern` might be enough to speed that up.

**Comment by foxfriends** [comment]
Closing in favour of #47
